ambry-network/src/main/java/com/github/ambry/network/http2/AmbrySendToHttp2Adaptor.java

```diff
@@ -71,9 +74,24 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
     }
     DefaultHttp2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(http2Headers, false);
     ctx.write(headersFrame);
-    ByteBuf dataContent = send.content();
-    DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(dataContent, true);
-    // Caller should call writeAndFlush().
-    ctx.write(dataFrame, promise);
+
+    // Referencing counting for derived {@link ByteBuf}: https://netty.io/wiki/reference-counted-objects.html#derived-buffers
+    try {
+      while (send.content().readableBytes() > maxFrameSize) {
+        ByteBuf slice = send.content().readSlice(maxFrameSize);
+        slice.retain();
+        DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(slice, false);
+        ctx.write(dataFrame, promise);
```

Comment:
it seems like we are sharing the same promise for every frame. Can that cause issues at other parts of the pipeline if called before all chunks have been written?

Reply:
good catch. removed it here. Only the last ctx.write() uses promise.